### PR TITLE
fix: clone newer tenants code to be used on users route

### PR DESCRIPTION
**Pull request by danielgent**
In tandem with this PR https://github.com/sesamyab/auth-admin/pull/49

### NOW
* I'm getting the integration tests working :heavy_check_mark: 
* I'm checking the PR diff and removing unneeded changes :heavy_check_mark: 

### NEXT
* I'll manually check the changes with React-admin :heavy_check_mark:  _working!_

### Future PRs
Check comments
* filtering
* sorting

Probably worth writing integration tests for these

### CODE REVIEW

File: src/adapters/planetscale/users/listUsers.ts [hunk 1]
```diff
@@ -3,72 +3,52 @@ import { ListUserParams } from "../../../adapters/interfaces/User";
...
-    tenantId,
-    { page, perPage, includeTotals, q }: ListUserParams,
-  ) => {
-    let query = db
-      .selectFrom("users")
-      .where("users.tenant_id", "=", tenantId)
-      .offset(page * perPage)
-      .limit(perPage);
+// duplicated from tenants
+function getCountAsInt(count: string | number | bigint) {
```
Comment (markusahlstrand) [inline src/adapters/planetscale/users/listUsers.ts:7]: Think we could remove this function and just use a parseInt instead

File: src/adapters/planetscale/users/listUsers.ts [hunk 2]
```diff
@@ -3,72 +3,52 @@ import { ListUserParams } from "../../../adapters/interfaces/User";
...
-              is_social: true,
-              connection: tag,
-            },
-          })),
-          ...userWithoutFields,
-          // Deprecated
-          tags,
-        };
+    if (params.q) {
+      query = query.where((eb) => eb.or([eb("name", "like", `%${params.q}%`)]));
```
Comment (markusahlstrand) [inline src/adapters/planetscale/users/listUsers.ts:26]: Maybe we want to search in some other fields here as well, like email?

**Review by markusahlstrand — APPROVED** [review]
Think this should work :)

**Comment by danielgent** [comment]
Nope! the issue is the `where tenantid` bit :smile: 

That's returning nothing

Probably our code should handle zero results but first I'll debug this using the planetscale console

**Comment by danielgent** [comment]
I'm making some progress I think as I'm developing locally (which is much much much easier than trying to debug from cloudflare)

But it seems like typescript isn't actually being checked.  I only see issues when I manually open the file

It seems like the tenant endpoint is working differently and returning results under a `tenant` key, but auth-admin is telling me to return the users under a `data` key

I don't know what the correct auth0 mgmt API spec is, so that's what I need to check next

**Comment by markusahlstrand** [comment]
I'm fairly certain that it is returned as an array under the "users" key.

It looks like that's how it's returned from auth2 today:
```
{
  "start": 0,
  "limit": 20,
  "length": "26074",
  "users": [
  ```

**Comment by markusahlstrand** [comment]
Ahh.. we need to move the users over to the "auth0DataProvider" in auth-admin. Then it should hopefully work..

**Comment by markusahlstrand** [comment]
Ok, think I understand the current state. I think it looks like it should work :)

File: src/adapters/in-memory/user/index.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ export function createUserAdapter(): UserDataAdapter {
   const users: UserResponse[] = [];
 
   return {
-    createUser: createUser(users),
-    listUsers: listUsers(users),
+    create: createUser(users),
+    list: listUsers(users),
```
Comment (danielgent) [inline src/adapters/in-memory/user/index.ts:12]: achieving consistency with the working `/api/v2/tenants` endpoint.

File: src/adapters/interfaces/User.ts [hunk 4]
```diff
@@ -3,15 +3,19 @@ import { PostUsersBody, UserResponse } from "../../types/auth0/UserResponse";
 
 export interface ListUserParams {
   page: number;
-  perPage: number;
-  includeTotals: boolean;
+  per_page: number;
+  include_totals: boolean;
```
Comment (danielgent) [inline src/adapters/interfaces/User.ts:7]: doesn't seem worth the effort to be changing between casing, especially given our build & lint process isn't picking up typescript changes.  We're mostly in vanilla JS land here :smile:

File: src/adapters/interfaces/User.ts [hunk 5]
```diff
@@ -3,15 +3,19 @@ import { PostUsersBody, UserResponse } from "../../types/auth0/UserResponse";
...
 export interface UserDataAdapter {
-  createUser(tenantId: string, PostUsersBody): Promise<UserResponse>;
-  listUsers(
+  create(tenantId: string, PostUsersBody): Promise<UserResponse>;
+  list(
     tenantId: string,
-    ListUserParams,
-  ): Promise<{ users: UserResponse[]; totals?: Totals }>;
+    params: ListUserParams,
+  ): Promise<{ data: UserResponse[]; totals?: Totals }>;
```
Comment (danielgent) [inline src/adapters/interfaces/User.ts:20]: I feel like this type should be defined once in a generic.

Comment (danielgent) [reply src/adapters/interfaces/User.ts:20]: ahhh, lol, this *IS* the generic interface :zipper_mouth_face:  :see_no_evil: 

Nice then, so the issue is that typescript isn't informing us of type violations

File: src/adapters/planetscale/users/listUsers.ts (hunk 1 above)
Comment (danielgent) [reply src/adapters/planetscale/users/listUsers.ts:7]: As mentioned on the other PR, I think VS code is correct here and our project isn't checking typescript properly.  I can prove this fairly easily

File: src/adapters/planetscale/users/listUsers.ts (hunk 2 above)
Comment (danielgent) [reply src/adapters/planetscale/users/listUsers.ts:26]: :shrug:  one step at a time I suppose!

File: src/adapters/planetscale/users/listUsers.ts [hunk 6]
```diff
@@ -3,72 +3,52 @@ import { ListUserParams } from "../../../adapters/interfaces/User";
...
+  }
+
+  return count;
+}
 
-    if (q) {
-      const filter = new URLSearchParams(q);
+export function listUsers(db: Kysely<Database>) {
+  return async (tenantId, params: ListUserParams) => {
+    let query = db.selectFrom("users").where("users.tenant_id", "=", tenantId);
```
Comment (danielgent) [inline src/adapters/planetscale/users/listUsers.ts:18]: this is all copying the newer code we have on the `listTenants` planetscale adaptor

Comment (danielgent) [reply src/adapters/planetscale/users/listUsers.ts:18]: Maybe only the initial `SELECT FROM` query is different, and the rest could be a single util function

File: src/adapters/planetscale/users/listUsers.ts [hunk 7]
```diff
@@ -3,72 +3,52 @@ import { ListUserParams } from "../../../adapters/interfaces/User";
...
 
-    return result;
+    const [{ count }] = await query
+      .select((eb) => eb.fn.countAll().as("count"))
+      .execute();
+
+    const countInt = getCountAsInt(count);
+
+    return {
+      data: users,
```
Comment (danielgent) [inline src/adapters/planetscale/users/listUsers.ts:48]: I was losing my mind here. Seems easier to always return the same shape as there are so many layers in this stack :smile:

File: integration-test/management-api/users.spec.ts [hunk 8]
```diff
@@ -1,14 +1,15 @@
 import { getAdminToken } from "../helpers/token";
 import { start } from "../start";
 
-describe("tenants", () => {
+describe("users", () => {
```
Comment (danielgent) [inline integration-test/management-api/users.spec.ts:4]: :smile: confusing!

Comment (danielgent) [reply integration-test/management-api/users.spec.ts:4]: Notice the rest of the test hasn't changed

File: integration-test/management-api/users.spec.ts [hunk 9]
```diff
@@ -1,14 +1,15 @@
...
   let worker;
   let token;
 
   beforeEach(async () => {
     worker = await start();
 
     token = await getAdminToken();
+    // should this not be /api/v2/tenants?
```
Comment (danielgent) [inline integration-test/management-api/users.spec.ts:12]: How does this work? :thinking:

File: src/adapters/in-memory/user/listUsers.ts [hunk 10]
```diff
@@ -4,10 +4,11 @@ import { ListUserParams } from "../../interfaces/User";
 export function listUsers(users: UserResponse[]) {
   return async (
     tenantId,
+    // why is our build process not picking up these types here? hard to spot issues...
```
Comment (danielgent) [inline src/adapters/in-memory/user/listUsers.ts:7]: this shows as an error in VS code. Would be amazing to have TypeScript really working :smile:

File: integration-test/management-api/users.spec.ts (hunk 9 above)
Comment (danielgent) [reply integration-test/management-api/users.spec.ts:12]: Seems like it can be both. Interesting

File: src/adapters/planetscale/users/index.ts [hunk 11]
```diff
@@ -8,7 +8,8 @@ export function createUsersAdapter(env: Env): UserDataAdapter {
   const db = getDb(env);
 
   return {
-    createUser: createUser(db),
-    listUsers: listUsers(db),
+    create: createUser(db),
+    // getting more typescript errors here but not being picked up anywhere... is this a VS code thing? is this a build issue?
```
Comment (danielgent) [inline src/adapters/planetscale/users/index.ts:12]: :eyes: this is probably something I've introduced.  This should really break the PR for such a critical project

File: src/routes/management-api/users.ts [hunk 12]
```diff
@@ -55,26 +55,20 @@ export class UsersMgmtController extends Controller {
...
-    const data = await env.data.users.listUsers(tenantId, {
+    const result = await env.data.users.list(tenantId, {
       page,
-      perPage: per_page,
-      includeTotals: include_totals,
+      per_page,
+      include_totals,
+      // TODO
+      // sort: parseSort(sort),
+      // q,
```
Comment (danielgent) [inline src/routes/management-api/users.ts:64]: I'll implement these as I keep playing around with auth-admin.  I could write integration tests for this... 
* create a new tenant
* push up some users `a@foo.com, b@foo.com`
* do sorting

IMHO not writing comprehensive tests is a time sink :smile:

File: src/routes/management-api/users.ts [hunk 13]
```diff
@@ -145,13 +139,10 @@ export class UsersMgmtController extends Controller {
...
 
-    const { id, ...data } = await env.data.users.createUser(tenantId, user);
+    const { id, ...data } = await env.data.users.create(tenantId, user);
 
     this.setStatus(201);
-    return {
-      ...data,
-      user_id: id,
-    };
+    return data;
```
Comment (danielgent) [inline src/routes/management-api/users.ts:145]: I've changed something here so I'm just returning the `user_id` field directly...

File: integration-test/management-api/users.spec.ts (hunk 9 above)
Comment (markusahlstrand) [reply integration-test/management-api/users.spec.ts:12]: Hmm.. think the old endpoint might still be around. We should remove that.

File: src/adapters/interfaces/User.ts (hunk 5 above)
Comment (markusahlstrand) [reply src/adapters/interfaces/User.ts:20]: So, do we want to pass an array called data here? It doesn't match with the rest-api, but.. it doesn't have to :) Maybe it's better to keep the data-adapter consistent?

File: src/adapters/planetscale/users/listUsers.ts (hunk 1 above)
Comment (markusahlstrand) [reply src/adapters/planetscale/users/listUsers.ts:7]: Ah, so parseInt doesn't take a number.. Guess parseInt(count.toString(), 10) would work?

File: src/routes/management-api/users.ts (hunk 13 above)
Comment (markusahlstrand) [reply src/routes/management-api/users.ts:145]: Ah, yes.. we should handle this in react-admin and be as true to the auth0 api as possible here.

File: src/types/auth0/UserResponse.ts [hunk 14]
```diff
@@ -40,5 +40,5 @@ export interface UserResponse extends BaseUser {
 }
 
 export interface GetUserResponseWithTotals extends Totals {
-  users: UserResponse[];
+  data: UserResponse[];
```
Comment (markusahlstrand) [inline src/types/auth0/UserResponse.ts:43]: This is not according to the auth0 api, right?

**Review by markusahlstrand — COMMENTED** [review]
Looks good. A few questions

File: src/adapters/planetscale/users/listUsers.ts (hunk 1 above)
Comment (danielgent) [reply src/adapters/planetscale/users/listUsers.ts:7]: Markus, I do not care about saving lines of code on this function :wink:

File: src/adapters/interfaces/User.ts (hunk 5 above)
Comment (danielgent) [reply src/adapters/interfaces/User.ts:20]: Cool cool, I'll change this back after lunch and then fix the auth-admin PR to be like it was (and check the auth0 specs) :+1: 

Indeed *everything needs to match the spec*!

File: integration-test/management-api/users.spec.ts (hunk 9 above)
Comment (danielgent) [reply integration-test/management-api/users.spec.ts]: I really don't see it though! If I search the code for `@Route("tenants` then I only find nested routes e.g. `@Route("tenants/{tenantId}/applications")`

It's not an issue so much as for my understanding :+1:

File: src/routes/management-api/users.ts (hunk 13 above)
Comment (danielgent) [reply src/routes/management-api/users.ts:145]: Indeed. The final test of all this will be getting `node-auth0` to work with this :sunglasses:

**Comment by danielgent** [comment]
TODO
* need to always return the users results under `users` if totals included!
*THEN* fix the auth-admin PR to handle this

File: integration-test/management-api/users.spec.ts (hunk 9 above)
Comment (markusahlstrand) [reply integration-test/management-api/users.spec.ts]: odd :)

**Comment by danielgent** [comment]
lol, I get an empty 204 after the last change (which was changing the keys in the response) :upside_down_face: 

Doesn't really make a lot of sense... something must be going wrong with the build

**Comment by danielgent** [comment]
I've no idea @markusahlstrand. I'll leave this PR for now and continue with the others

It works in react-admin :shrug: 

I don't see why changing a key would cause a 204, I can only assume it's not building *but* we don't get informed by typescript unless we open the file :upside_down_face:

**Pull request merged**